src/cascade/model/grids.py

```diff
@@ -0,0 +1,234 @@
+        """Construct an age-time grid with uniform spacing on each axis.
+        All values are in units of years.
+
+        Args:
+            age_start: the lowest age in the grid
+            age_end: the highest age in the grid
+            age_step: the spacing of age points
+            time_start: the earliest time in the grid
+            time_end: the latest time in the grid
+            time_step: the spacing of time points
+        """
+        ages = np.arange(age_start, age_end, age_step)
+        times = np.arange(time_start, time_end, time_step)
+
+        return cls(ages, times)
+
+    def __init__(self, ages, times, snap_distance=GRID_SNAP_DISTANCE):
+        self._ages = tuple(sorted(set(ages)))
+        self._times = tuple(sorted(set(times)))
+        self._snap_distance = snap_distance
+
+    @property
+    def ages(self):
+        return self._ages
+
+    @property
+    def times(self):
+        return self._times
+
+    @property
+    def snap_distance(self):
+        return self._snap_distance
+
+    def to_age(self, query):
+        if np.isinf(query):
+            return query
+        return _nearest(self._ages, query)
+
+    def to_time(self, query):
+        if np.isinf(query):
+            return query
+        return _nearest(self._times, query)
+
+    def __hash__(self):
+        return hash((self._ages, self._times, self._snap_distance))
+
+    def __eq__(self, other):
+        return (
+            isinstance(other, type(self))
+            and self._ages == other._ages
+            and self._times == other._times
+            and self._snap_distance == other._snap_distance
+        )
+
+
+def _any_close(value, targets, tolerance):
+    """True if any element of targets is within tolerance of value
+    """
+    for target in targets:
+        if isclose(value, target, abs_tol=tolerance):
+            return True
+    return False
+
+
+def _validate_region(grid, lower_age, upper_age, lower_time, upper_time):
+    """Validates that the described rectangle aligns with the grid.
+    Positive or negative infinite bounds will be treated as stretching to the edge of the grid.
+
+    Raises:
+        ValueError: If the rectangle does not align
+    """
+    if not _any_close(lower_age, grid.ages, grid.snap_distance) and not isinf(lower_age):
+        raise ValueError("Lower age not in underlying grid")
+
+    if not _any_close(upper_age, grid.ages, grid.snap_distance) and not isinf(upper_age):
+        raise ValueError("Upper age not in underlying grid")
+
+    if not _any_close(lower_time, grid.times, grid.snap_distance) and not isinf(lower_time):
+        raise ValueError("Lower time not in underlying grid")
+
+    if not _any_close(upper_time, grid.times, grid.snap_distance) and not isinf(upper_time):
+        raise ValueError("Upper time not in underlying grid")
+
+
+class _RegionView:
+    """Represents a slice into a PriorGrid and exposes a way to query or set
+    the prior over that slice.
```

Comment:
Based on the conversation I had with Greg, in which he had to explain this to me several times because I was making the same assumption as you are, that is not correct. points in the smoothing grid represent the prior from that point forward (in both age and time) until the next prior, if there is one. That's why I chose to represent priors as rectangles. Now it's possible that I'm still misunderstanding what actually happens in which case you are correct and this doesn't make any sense.

Reply:
Interpolation (in prior space) is constant, not bilinear is what I mean to say.